File: aspnet/src/Observability/EventLog.cs
namespace src.Observability;

public static class EventLog
{
    public static class Stage
    {
        public const string Kestrel = "kestrel";
        public const string Controller = "controller";
        public const string Service = "service";
        public const string AsyncOp = "async_op";
        public const string Db = "db";
    }

    public static void Log(string req, string stage, string @event, string? detail = null)
    {
        var ts = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var line = detail is null
            ? $"ts={ts} req={req} stage={stage} event={@event}"
            : $"ts={ts} req={req} stage={stage} event={@event} detail={detail}";
        Console.WriteLine(line);
    }
}

File: spring-webflux/src/main/java/com/example/webflux/observability/EventLog.java
package com.example.webflux.observability;

public final class EventLog {

    public static final class Stage {
        public static final String NETTY = "netty";
        public static final String CONTROLLER = "controller";
        public static final String SERVICE = "service";
        public static final String ASYNC_OP = "async_op";
        public static final String DB = "db";

        private Stage() {}
    }

    public static void log(String req, String stage, String event) {
        log(req, stage, event, null);
    }

    public static void log(String req, String stage, String event, String detail) {
        long ts = System.currentTimeMillis();
        String line = detail == null
                ? "ts=" + ts + " req=" + req + " stage=" + stage + " event=" + event
                : "ts=" + ts + " req=" + req + " stage=" + stage + " event=" + event + " detail=" + detail;
        System.out.println(line);
    }

    private EventLog() {}
}
